### DISCUSSION TITLE
Undocumented API Endpoints

### DISCUSSION THREAD

**Discussion by ssnider21** [joeyagreco/sleeper#11] [Ideas]
There's a decent number of api endpoints that aren't included in sleeper's docs, particularly the stats endpoints. Do you have any interest in adding these? I can work on adding some if so. Would be beneficial to both this library and leeger in my opinion.

Examples:
- `/stats/nfl/player/{player_id}?season_type=regular&season={year}&week={week_number}` to get a player's stats for a single week
- `/stats/nfl/{year}/{week_number}?season_type=regular&position[]=DEF&position[]=K&position[]=QB&position[]=RB&position[]=TE&position[]=WR` to get all stats for a week
- can also replace `stats` in both above endpoints with `projections`

**Comment by joeyagreco** [comment]
What's the full URL you are using for these?

Tried `https://api.sleeper.app/v1/stats/nfl/player/4039?season_type=regular&season=2021&week=9` to get 2021 week 9 stats for Cooper Kupp but the response seems to be rankings for each player.

**Comment by ssnider21** [comment]
Interesting, I was just using `https://api.sleeper.com` as the base url, without `/v1/`. This should return you just Kupp from week 9, 2021: `https://api.sleeper.com/stats/nfl/player/4039?season_type=regular&season=2021&week=9`. You can use `.app` or `.com`, it seems to make no difference.

**Reply by joeyagreco** [reply]
Now that I think about it, the reason Sleeper probably doesn't have these in their official documentation is because they aren't meant to be used by consumers.

Looks like they are getting this data from other sites which is noted in the `company` field in the responses (sportradar, rotowire, etc...)

These endpoints will likely never have official public support from Sleeper, and should be considered unstable since they could be blocked/deprecated at any point.

I'll still add support for them to this library but I'll put them in a separate API Client for unofficial/volatile endpoints.

**Reply by ssnider21** [reply]
Yeah I agree, that's most likely the case. That's a good approach for sure.

**Comment by joeyagreco** [comment]
Weird, but good find!

I wonder how many endpoints are hidden outside of the `v1` API.

I can/will add supports for these endpoints soon.. thank you!

**Comment by joeyagreco** [comment]
TODO
___
- [x] Create new "Unofficial" folder to hold API Clients for hidden endpoints, noting that support for these is not guaranteed by Sleeper
- [x] Add support for player stats with filters
- [x] Add support for single player combined season stats
- [x] Add support for week stats with filters
- [x] Add support for player projections with filters
- [x] Add support for single player combined season projections
- [x] Add support for week projections with filters
- [x] Add support for team depth chart
- [x] Add support for player headshot
- [x] Add support for year schedule
- [x] Write tests

**Comment by ssnider21** [comment]
I'll start a working list of undocumented endpoints I've found here. All will use base url `https://api.sleeper.com` unless otherwise noted - I'm assuming this is maybe the internal sleeper api endpoint, but it *is* publicly available as read-only. 

Roster Related:
- Team Depth Chart: `players/nfl/{team_abbreviation}/depth_chart`
- Player Head Shot: `https://sleepercdn.com/content/nfl/players/{player_id}.jpg`

Stats/Projections:
- Single player week stats/projections: `/{stats|projections}/nfl/player/{player_id}?season_type=regular&season={year}&week={week_number}`
- All players week stats/projections: `/{stats|projections}/nfl/{year}/{week_number}?season_type=regular&position[]=DEF&position[]=K&position[]=QB&position[]=RB&position[]=TE&position[]=WR` any combination of query params for positions is acceptable here as well
- Single Player combined season stats/projections: `/{stats|projections}/nfl/player/{player_id}?season_type=regular&season={year}`

NFL Related:
- Year Schedule: `schedule/nfl/regular/{year}`

There's also a graphql endpoint that most of the requests on the sleeper website are made through - that might be outside the scope of this library though.

I'm happy to help out on any of this work, just let me know!

**Reply by joeyagreco** [reply]
Great research! Much appreciated.

I don't see an issue with adding support to any of these.

As for the graphql endpoint/s, I think those can be used in this project, but I'll probably add support for those last as they'll likely take a bit longer to set up.

**Reply by ssnider21** [reply]
Sounds good. Graphql example in that case, using Kupp again:
endpoint: ` https://sleeper.com/graphql`
query: 
```graphql
query get_player_score_and_projections_batch {
        
                  nfl__regular__2021__1__stat: stats_for_players_in_week(sport: "nfl",season: "2021",category: "stat",season_type: "regular",week: 1,player_ids: ["4039"]){
                    game_id
                    opponent
                    player_id
                    stats
                    team
                    week
                    season
                            }
                
          
                  nfl__regular__2021__1__proj: stats_for_players_in_week(sport: "nfl",season: "2021",category: "proj",season_type: "regular",week: 1,player_ids: ["4039"]){
                    game_id
                    opponent
                    player_id
                    stats
                    team
                    week
                    season
                            }
                          
                     }
```

I'll see if I can find what other graphql operations are available.

**Comment by joeyagreco** [comment]
@ssnider21 
[Here's](https://github.com/joeyagreco/sleeper/pull/12) the pull request to add these features.

Please let me know what you think:
- Any potential bugs you see
- Any features I missed
- Any suggestions

**Reply by ssnider21** [reply]
Hey sorry for the slow review, but I got a chance to play around with the branch last night. Didn't encounter any bugs! Honestly I don't really have any suggestions, this looks like it covers all the hidden endpoints I was able to find.
